File: src/newinterpreter.js
const {isFunction, isObject, isString, isArray} = require("../src/type-utils");

class Interpreter {
    constructor(context) {
        this.context = context;
    }

    visit(node) {
        let funcName = "visit_" + node.constructor.name;
        return this[funcName](node);
    }

    visit_ASTNode(node) {
        let str;
        switch (node.token.kind) {
            case("number"):
                return +node.token.value;
            case("null"):
                return null;
            case("string"):
                str = node.token.value.slice(1, -1);
                return str;
            case("true"):
                return true;
            case("false"):
                return false;
        }
    }

    visit_UnaryOp(node) {
        switch (node.token.kind) {
            case ("+"):
                return +this.visit(node.expr);
            case ("-"):
                return -this.visit(node.expr);
            case ("!"):
                return !this.visit(node.expr);
        }
    }

    visit_BinOp(node) {
        switch (node.token.kind) {
            case ("+"):
                return this.visit(node.left) + this.visit(node.right);
            case ("-"):
                return this.visit(node.left) - this.visit(node.right);
            case ("/"):
                return this.visit(node.left) / this.visit(node.right);
            case ("*"):
                return this.visit(node.left) * this.visit(node.right);
            case (">"):
                return this.visit(node.left) > this.visit(node.right);
            case ("<"):
                return this.visit(node.left) < this.visit(node.right);
            case (">="):
                return this.visit(node.left) >= this.visit(node.right);
            case ("<="):
                return this.visit(node.left) <= this.visit(node.right);
            case ("!="):
                return this.visit(node.left) != this.visit(node.right);
            case ("=="):
                return this.visit(node.left) == this.visit(node.right);
            case ("||"):
                return this.visit(node.left) || this.visit(node.right);
            case ("&&"):
                return this.visit(node.left) && this.visit(node.right);
            case ("**"):
                return Math.pow(this.visit(node.left), this.visit(node.right));
            case ("."): {
                let obj = this.visit(node.left);
                let key = node.right.token.value;

                if (obj.hasOwnProperty(key)) {
                    return obj[key];
                }
                break
            }
            case ("in"): {
                let left = this.visit(node.left);
                let right = this.visit(node.right);

                if (isObject(right)) {
                    right = Object.keys(right);
                } else if (isString(right)) {
                    return right.indexOf(left) !== -1;
                }

                return right.some(r => isEqual(left, r));
            }
        }
    }

    visit_List(node) {
        let list = [];

        if (node.list[0] != undefined) {
            node.list.forEach(function (item) {
                list.push(this.visit(item))
            }, this);
        }

        return list
    }

    visit_ArrayAccess(node) {
        let array = this.context[node.token.value];
        let left = null, right = null;

        if (node.left) {
            left = this.visit(node.left);
        }
        if (node.right) {
            right = this.visit(node.right);
        }
        if (left < 0) {
            left = array.length + left
        }
        if (node.isInterval) {
            right = right === null ? array.length : right;
            if (right < 0) {
                right = array.length + right;
                if (right < 0)
                    right = 0
            }
            if (left > right) {
                left = right
            }
            return array.slice(left, right)
        } else {
            return array[left]
        }
    }

    visit_Builtin(node) {
        let args = [];
        let builtin = this.context[node.token.value];
        if (isFunction(builtin)) {
            node.args.forEach(function (item) {
                args.push(this.visit(item))
            }, this);
            return builtin.apply(null, args);
        }
        return builtin
    }

    visit_Object(node) {
        let obj = {};

        for (let key in node.obj) {
            obj[key] = this.visit(node.obj[key])
        }

        return obj
    }

    interpret(tree) {
        return this.visit(tree);
    }
}

let isEqual = (a, b) => {
    if (isArray(a) && isArray(b) && a.length === b.length) {
        for (let i = 0; i < a.length; i++) {
            if (!isEqual(a[i], b[i])) {
                return false;
            }
        }
        return true;
    }
    if (isFunction(a)) {
        return a === b;
    }
    if (isObject(a) && isObject(b)) {
        let keys = Object.keys(a).sort();
        if (!isEqual(keys, Object.keys(b).sort())) {
            return false;
        }
        for (let k of keys) {
            if (!isEqual(a[k], b[k])) {
                return false;
            }
        }
        return true;
    }
    return a === b;
};

exports
    .NewInterpreter = Interpreter;

File: jsone/newinterpreter.py
from .shared import TemplateError, InterpreterError, string
import json


def infixExpectationError(operator, expected):
    return InterpreterError('infix: {} expects {} {} {}'.
                            format(operator, expected, operator, expected))


class Interpreter:
    def __init__(self, context):
        self.context = context

    def visit(self, node):
        method_name = 'visit_' + type(node).__name__
        visitor = getattr(self, method_name)
        return visitor(node)

    def visit_ASTNode(self, node):
        if node.token.kind == "number":
            v = node.token.value
            return float(v) if '.' in v else int(v)
        elif node.token.kind == "null":
            return None
        elif node.token.kind == "string":
            return node.token.value[1:-1]
        elif node.token.kind == "true":
            return True
        elif node.token.kind == "false":
            return False

    def visit_UnaryOp(self, node):
        value = self.visit(node.expr)
        if node.token.kind == "+":
            if not isNumber(value):
                raise InterpreterError('{} expects {}'.format('unary +', 'number'))
            return value
        elif node.token.kind == "-":
            if not isNumber(value):
                raise InterpreterError('{} expects {}'.format('unary -', 'number'))
            return -value
        elif node.token.kind == "!":
            return not self.visit(node.expr)

    def visit_BinOp(self, node):
        left = self.visit(node.left)
        if node.token.kind == "||":
            return bool(left or self.visit(node.right))
        elif node.token.kind == "&&":
            return bool(left and self.visit(node.right))
        elif node.token.kind == ".":
            right = node.right.value
        else:
            right = self.visit(node.right)

        if node.token.kind == "+":
            if not isinstance(left, (string, int, float)) or isinstance(left, bool):
                raise infixExpectationError('+', 'number/string')
            if not isinstance(right, (string, int, float)) or isinstance(right, bool):
                raise infixExpectationError('+', 'number/string')
            if type(right) != type(left) and \
                    (isinstance(left, string) or isinstance(right, string)):
                raise infixExpectationError('+', 'numbers/strings')
            return left + right
        elif node.token.kind == "-":
            testMathOperands("-", left, right)
            return left - right
        elif node.token.kind == "/":
            testMathOperands("/", left, right)
            return left / right
        elif node.token.kind == "*":
            testMathOperands("*", left, right)
            return left * right
        elif node.token.kind == ">":
            testComparisonOperands(">", left, right)
            return left > right
        elif node.token.kind == "<":
            testComparisonOperands("<", left, right)
            return left < right
        elif node.token.kind == ">=":
            testComparisonOperands(">=", left, right)
            return left >= right
        elif node.token.kind == "<=":
            testComparisonOperands("<=", left, right)
            return left <= right
        elif node.token.kind == "!=":
            return left != right
        elif node.token.kind == "==":
            return left == right
        elif node.token.kind == "**":
            testMathOperands("**", left, right)
            return right ** left
        elif node.token.value == "in":
            if isinstance(right, dict):
                if not isinstance(left, string):
                    raise infixExpectationError('in-object', 'string on left side')
            elif isinstance(right, string):
                if not isinstance(left, string):
                    raise infixExpectationError('in-string', 'string on left side')
            elif not isinstance(right, list):
                raise infixExpectationError(
                    'in', 'Array, string, or object on right side')
            try:
                return left in right
            except TypeError:
                raise infixExpectationError('in', 'scalar value, collection')

        elif node.token.kind == ".":
            if not isinstance(left, dict):
                raise infixExpectationError('.', 'object')
            try:
                return left[right]
            except KeyError:
                raise TemplateError(
                    '"{}" not found in {}'.format(right, json.dumps(left)))

    def visit_List(self, node):
        list = []

        if node.list[0] is not None:
            for item in node.list:
                list.append(self.visit(item))

        return list

    def visit_ValueAccess(self, node):
        value = self.visit(node.arr)
        left = 0
        right = None

        if node.left:
            left = self.visit(node.left)
        if node.right:
            right = self.visit(node.right)

        if isinstance(value, (list, string)):
            if node.isInterval:
                if right is None:
                    right = len(value)
                try:
                    return value[left:right]
                except TypeError:
                    raise InterpreterError('should only use integers to access arrays or strings')
            else:
                try:
                    return value[left]
                except IndexError:
                    raise TemplateError('index out of bounds')
                except TypeError:
                    raise InterpreterError('should only use integers to access arrays or strings')

        if not isinstance(value, dict):
            raise infixExpectationError('[..]', 'object, array, or string')
        if not isinstance(left, string):
            raise infixExpectationError('[..]', 'string index')

        try:
            return value[left]
        except KeyError:
            return None

    def visit_Builtin(self, node):
        args = []
        try:
            builtin = self.context[node.token.value]
        except KeyError:
            raise InterpreterError(
                'unknown context value {}'.format(node.token.value))
        if callable(builtin):
            if node.args is not None:
                for item in node.args:
                    args.append(self.visit(item))
                return builtin(*args)
        return builtin

    def visit_Object(self, node):
        obj = {}
        for key in node.obj:
            obj[key] = self.visit(node.obj[key])
        return obj

    def interpret(self, tree):
        return self.visit(tree)


def testMathOperands(op, left, right):
    if not isNumber(left):
        raise infixExpectationError(op, 'number')
    if not isNumber(right):
        raise infixExpectationError(op, 'number')
    return


def testComparisonOperands(op, left, right):
    if type(left) != type(right) or \
            not (isinstance(left, (int, float, string)) and not isinstance(left, bool)):
        raise infixExpectationError(op, 'numbers/strings')
    return


def isNumber(v):
    return isinstance(v, (int, float)) and not isinstance(v, bool)
